### DISCUSSION TITLE
Handling second order element

### DISCUSSION THREAD

**Discussion by Clemoreau** [nschloe/meshio#1202] [Q&A]
Hello all, 

I was wondering if meshio could handle second order elements (especially with .fem format). Because I read a mesh containing second order elements and the result is not what I expected : something seems not be working in the node's connectivity.
![image](https://user-images.githubusercontent.com/52700851/137325361-5f73a6c3-4aed-4571-981f-e1474c1c9772.png)

I don't know if it's a bug or if it's something which was not foreseen in the development. I could eventually contribute to the repository if necessary.

Thank you

**Comment by nschloe** [comment]
Should work, it's a bug if it doesn't. It looks like something with the node ordering is off...

**Comment by Clemoreau** [comment]
Ok, it's a bug then, do you want me to continue the discussion here or to open an issue ? 

When I import one CHEXA element (second order -> 20 nodes), I obtain that :

![image](https://user-images.githubusercontent.com/52700851/137339380-f7710b0e-147e-44fe-9a11-d316ae641cee.png)

I think you're right, the problem is in the node ordering

**Reply by nschloe** [reply]
The best thing would bit a PR straight away. :sunglasses:

**Comment by Clemoreau** [comment]
Ahah yes I guess so, I'll look at it, 

Can you confirm that I'm looking at the right place in _nastran.py / function read_buffer() ? (it is for .fem file)

**Reply by nschloe** [reply]
Sure. You'll probably have to adapt the write function as well to get the tests to pass.

**Reply by Clemoreau** [reply]
Do you have any reference about the node ordering for the second order CHEXA (hexahedron20) ? I'm a bit lost about it. Currently it's [0,1,2,3,...,17,18,19]

![image](https://user-images.githubusercontent.com/52700851/137713904-2cb58464-09c3-471f-bd60-9aacca9c1259.png)

**Reply by nschloe** [reply]
meshio uses VTK ordering for everything, see [here](https://vtk.org/wp-content/uploads/2015/04/file-formats.pdf):

![screenshot](https://user-images.githubusercontent.com/181628/137714898-ffed7577-5e55-4abb-b479-35a99125ea0d.png)

**Reply by Clemoreau** [reply]
Thank you for the reference ! The problem was effectively in the node ordering for hexahedron20 and wedge15. Now it's fixed.

I would like to add a parameter 'is_quadratic' to the write/read function to write/read a linear mesh from a quadratic mesh if False. Do you agree if I add that to my PR ? (default value would be True). Do you agree ?
The matter is that I often have to deal with quadratic mesh while I don't need such a level of detail and it's easier to work with linear mesh. It's a kind of mesh simplification.


Thank you !

**Reply by nschloe** [reply]
> I would like to add a parameter 'is_quadratic'

That'd just complicate things. What we could do instead is add a command-line too, something like
```
meshio force-order 2 infile outfile
```
which forces all cells in `infile` to have a certain order. This could be used to reduce all cells to order 1, or augment all cells.

**Reply by Clemoreau** [reply]
Ok, I think that it's a bit too complicated for me, but I will consider it if I realise that it is useful in my work. It was just an idea but I'm not sure of how doing it.

Still, I corrected the ordering of the nodes for hexahedron20 and wedge15 on my fork and I would like to add these modifications to the main branch of meshio if you agree.

I create a PR for this purpose.

Thank you for your kind help :)

**Reply by nschloe** [reply]
> Ok, I think that it's a bit too complicated for me

Shouldn't be too hard I think. Anyway, I appreciate the effort!
